idea-plugin/src/main/kotlin/arrow/meta/ide/plugins/proofs/folding/kinds.kt

```diff
@@ -1,58 +1,43 @@
 package arrow.meta.ide.plugins.proofs.folding
 
-import arrow.meta.ide.dsl.utils.typeProjections
-import arrow.meta.ide.dsl.utils.typeReferences
-import com.intellij.lang.ASTNode
-import com.intellij.lang.folding.FoldingBuilderEx
-import com.intellij.lang.folding.FoldingDescriptor
-import com.intellij.openapi.editor.Document
+import arrow.meta.ide.IdeMetaPlugin
+import arrow.meta.phases.ExtensionPhase
 import com.intellij.psi.PsiElement
-import com.intellij.psi.impl.source.SourceTreeToPsiMap
-import org.jetbrains.kotlin.psi.KtElement
+import com.intellij.psi.util.strictParents
+import org.jetbrains.kotlin.psi.KtTypeProjection
+import org.jetbrains.kotlin.psi.KtTypeReference
+import org.jetbrains.kotlin.psi.KtUserType
+import org.jetbrains.kotlin.resolve.descriptorUtil.fqNameSafe
+import org.jetbrains.kotlin.types.KotlinType
+import org.jetbrains.kotlin.utils.addToStdlib.safeAs
 
-class KindFoldingBuilder : FoldingBuilderEx() {
-  override fun getPlaceholderText(node: ASTNode): String? =
-    SourceTreeToPsiMap.treeElementToPsi(node)?.let { psiElement ->
-      (psiElement as KtElement).typeProjections
-        .filter { !it.text.startsWith("Kind") }
-        .map { it.text }
-        .toString()
-        .replace("[", "")
-        .replaceFirst(", ", "<")
-        .replace("]", ">")
+val IdeMetaPlugin.codeFoldingOnKinds: ExtensionPhase
+  get() = addFoldingBuilder(
+    isTypeMatching = ::parentTypeMatches,
+    toFoldString = KtTypeReference::foldString
+  )
+
+private fun parentTypeMatches(typeReference: KtTypeReference): Boolean =
+  typeReference.getType().isTypeMatching() &&
+    typeReference.strictParents().all { psiElement ->
+      !psiElement.safeAs<KtTypeReference>()?.getType().isTypeMatching()
     }
 
-  override fun buildFoldRegions(root: PsiElement, document: Document, quick: Boolean): Array<FoldingDescriptor> =
-    (root as KtElement).typeReferences
-      .filter { it.text.startsWith("Kind") }
-      .map { FoldingDescriptor(it, it.textRange) }
-      .toTypedArray()
+private fun KotlinType?.isTypeMatching() =
+  this?.constructor?.declarationDescriptor?.fqNameSafe?.asString() == "arrowx.Kind22"
 
-  override fun isCollapsedByDefault(node: ASTNode): Boolean =
-    true
+private fun KtTypeReference.foldString(): String {
+  val children: Array<out PsiElement> = firstChild.safeAs<KtUserType>()?.typeArgumentList?.children.orEmpty()
+  return children[0].text + children.toList().subList(1, children.size).joinToString(
```

Comment:
this will throw an error, if the array is empty. Try to encode this without unsafe access